compiler/src/iree/compiler/Codegen/Common/GPU/AMDGPUDistributeContract.cpp

```diff
@@ -0,0 +1,269 @@
+          contractOp, "missing nested layout for contraction result");
+    }
+    int64_t rank = resultLayout.getBatchOrder().size();
+
+    NestedLayoutAttr lhsLayout =
+        dyn_cast<NestedLayoutAttr>(signature[contractOp.getLhs()]);
+    if (!lhsLayout) {
+      return rewriter.notifyMatchFailure(
+          contractOp, "missing nested layout for contraction lhs");
+    }
+    NestedLayoutAttr rhsLayout =
+        dyn_cast<NestedLayoutAttr>(signature[contractOp.getRhs()]);
+    if (!rhsLayout) {
+      return rewriter.notifyMatchFailure(
+          contractOp, "missing nested layout for contraction rhs");
+    }
+
+    auto mfmaAttr =
+        contractOp->getAttrOfType<IREE::GPU::MFMAAttr>("iree.amdgpu.mfma");
+    if (!mfmaAttr) {
+      return rewriter.notifyMatchFailure(
+          contractOp, "missing iree.amdgpu.mfma intrinsic attribute");
+    }
+    auto [aVectorType, bVectorType, cVectorType] = mfmaAttr.getABCVectorTypes();
+
+    ContractKind contractKind =
+        inferContractKind(getContext(), contractOp.getIndexingMapsArray());
+    if (contractKind == ContractKind::UNKNOWN) {
+      return rewriter.notifyMatchFailure(contractOp, "unknown contract kind");
+    }
+
+    SmallVector<int64_t> distShape = resultLayout.getDistributedShape();
+    LLVM_DEBUG({
+      llvm::dbgs() << "distributed shape: [";
+      llvm::interleaveComma(distShape, llvm::dbgs());
+      llvm::dbgs() << "]\n";
+    });
+
+    // Create a zero vector with the full distributed vector shape for
+    // accumulating unrolled contraction results.
+    auto tileType = VectorType::get(distShape, resultType.getElementType());
+    Value zero = rewriter.create<arith::ConstantOp>(
+        contractOp.getLoc(), tileType, rewriter.getZeroAttr(tileType));
+    VectorValue finalTile = cast<VectorValue>(zero);
+    LLVM_DEBUG(llvm::dbgs() << "init tile: " << finalTile << "\n");
+
+    SmallVector<int64_t, 2> lhsBatchOffsets(rank, 0);
+    SmallVector<int64_t, 2> rhsBatchOffsets(rank, 0);
+
+    ArrayRef<int64_t> resultBatches = resultLayout.getBatchesPerSubgroup();
+    SmallVector<int64_t, 2> resultBatchTileSizes(rank, 1);
+    LLVM_DEBUG({
+      llvm::dbgs() << "result batches: [";
+      llvm::interleaveComma(resultBatches, llvm::dbgs());
+      llvm::dbgs() << "]\n";
+    });
+
+    // Iterate over all result batches and unroll computation to direct MFMA
+    // intrinsic ops.
+    Location loc = contractOp.getLoc();
+    auto resultTiles = StaticTileOffsetRange(
+        resultBatches, resultBatchTileSizes, resultLayout.getBatchOrder());
+    for (SmallVector<int64_t, 2> resultBatchOffsets : resultTiles) {
+      LLVM_DEBUG({
+        llvm::dbgs() << "current result batch offsets: [";
+        llvm::interleaveComma(resultBatchOffsets, llvm::dbgs());
+        llvm::dbgs() << "]\n";
+      });
+
+      // Get the slice of the accumulator in this batch.
+      Value acc = getDistributed(
+          rewriter, cast<VectorValue>(contractOp.getAcc()), resultLayout);
```

Comment:
nit: I think `getDistributed` can create `to_simt` operations, so best to call it one outside the loop.

edit: actually I think the conversion gets cached

Reply:
Yeah it gets cached. But still good point of moving it out.